### PR TITLE
Add ErrorOr class

### PR DESCRIPTION
**Pull request by MrCreosote**
This class will be used to return either an object, such as a Workspace
object, or an error code that explains why the object could not be
retrieved. The error code will be exposed in API methods that use the
ignoreError parameter rather than just a null value.

### CODE REVIEW

**Comment by MrCreosote** [comment]
ugh, apt update errors

**Comment by MrCreosote** [comment]
@jsfillman Looks like LGTM isn't working correclty here, it's not detecting the java code

File: build.xml [hunk 1]
```diff
@@ -62,7 +62,6 @@
     <include name="apache_commons/commons-io-2.4.jar"/>
     <include name="apache_commons/commons-lang3-3.1.jar"/>
     <include name="mongo/mongo-java-driver-3.8.2.jar"/>
-    <include name="bson4jackson/bson4jackson-2.2.0-2.2.0.jar"/>
```
Comment (xpe) [inline build.xml:65]: Did this sneak in from another cause/change? (That's my guess)

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.Optional;
+
+/** 
+ * A class that represents an object or an error that occurred attempting to create or retrieve
+ * the object.
+ */
+public class ErrorOr<T> {
+
+	// TODO ERROR may want to support including custom error fields? Different errors = different fields
+	// TODO LOMBOK
```
Comment (xpe) [inline src/us/kbase/workspace/database/exceptions/ErrorOr.java:15]: Lombok = https://projectlombok.org/ ?

**Comment by xpe** [comment]
From Gavin:

> There are workspace methods that return multiple things. By default if getting one thing fails (authorization, deleted, no such id etc) the whole call fails. You can ask for partial results but the things are just missing from the results. This is step 1 of many of also returning why you can't get the thing, which is basically required when doing integrations with other systems, which currently are a lot harder than they need to be

File: src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java [hunk 3]
```diff
@@ -1014,6 +1014,7 @@ public ResolvedWorkspaceID resolveWorkspace(
 			final Set<WorkspaceIdentifier> wsis,
 			final boolean allowDeleted,
 			final boolean allowMissing)
+			// TODO CODE don't think comm exception needs to be checked. Usually will just fail.
```
Comment (xpe) [inline src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java:1017]: Can you add a test for this?

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	// TODO CODE return this from bulk methods rather than object / null 
+	
+	private final ErrorType error;
```
Comment (xpe) [inline src/us/kbase/workspace/database/exceptions/ErrorOr.java:18]: I can see that coupling `ErrorType` with `ErrorTo` here might be pragmatic... however, I could see `ErrorOr` becoming something we use more broadly, in which case we'd want the error type to be a second generic.

File: build.xml (hunk 1 above)
Comment (MrCreosote) [reply build.xml:65]: dependency tree cruft that was missed when removing the root jar in the tree

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java (hunk 2 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/exceptions/ErrorOr.java:15]: correct

File: src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java (hunk 3 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java:1017]: Not sure what you mean, this is just a comment. There's nothing to test.

Maybe you mean changing the communication exception to an unchecked exception when and if that happens? There's no way to test for that either, though...

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java (hunk 4 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/exceptions/ErrorOr.java:18]: That could make sense, but in the case of the workspace not necessary, so I'm pulling a YAGNI here.

File: build.xml (hunk 1 above)
Comment (MrCreosote) [reply build.xml:65]: Really need to move to gradle at some point

File: src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java (hunk 3 above)
Comment (xpe) [reply src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java:1017]: > TODO CODE don't think comm exception needs to be checked. Usually will just fail.

When I see a TODO comment, I try to ask "What is actionable here?" with the hopes of moving it to a ticketing system. For this one, I'd also ask how we can more certainty around the "I don't think" part.

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java (hunk 4 above)
Comment (xpe) [reply src/us/kbase/workspace/database/exceptions/ErrorOr.java:18]: Yes, we are saying the same thing in this case.

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java (hunk 2 above)
Comment (xpe) [reply src/us/kbase/workspace/database/exceptions/ErrorOr.java:15]: I'd like to move TODOs out of the code into tickets.

File: build.xml (hunk 1 above)
Comment (xpe) [reply build.xml:65]: 1. If you get a chance, splitting this change out into its own commit helps.

2. What's the next step re: Gradle: create a ticket?

Comment (MrCreosote) [reply build.xml:65]: For 1): yeaaaah... I'm not going to spend my time on that. Sorry.
For 2): Nothing, I'll get to it when I get to it. It's not something that needs to be tracked.

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java (hunk 2 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/exceptions/ErrorOr.java:15]: It seems you and I are pretty far apart on how we treat TODOs. I'm happy to have that discussion, but not in a PR review - we should probably move it to slack or something.

File: src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java (hunk 3 above)
Comment (MrCreosote) [reply src/us/kbase/workspace/database/mongo/MongoWorkspaceDB.java:1017]: See above re TODOs. Re certainty, I need to look at how the exception is used in the code base, which is not something I'm going to get to immediately and is not important enough for a ticket.

File: src/us/kbase/workspace/database/exceptions/ErrorOr.java (hunk 2 above)
Comment (xpe) [reply src/us/kbase/workspace/database/exceptions/ErrorOr.java:15]: We can discuss over a video chat.

**Comment by MrCreosote** [comment]
There haven't been any new comments in several days, so I'm going to go ahead and merge.

**Comment by xpe** [comment]
I didn't approve the PR. You can find someone else if you want an approval.

**Pull request merged**